components/core/src/clp/regex_utils/regex_translation_utils.cpp

```diff
@@ -0,0 +1,245 @@
+            case TranslatorState::RegexPatternState::END:
+                ec = end_state_transition(state, it, wildcard_str, config);
+                break;
+            default:
+                ec = ErrorCode::IllegalState;
+                break;
+        }
+        if (ec) {
+            return ec;
+        }
+        ++it;
+    }
+
+    ec = final_state_cleanup(state, it, wildcard_str, config);
+    if (ec) {
+        return ec;
+    }
+    return wildcard_str;
+}
+
+namespace {
+/**
+ * Treats each character literally and directly append it to the wildcard string, unless it is a
+ * meta-character.
+ *
+ * Each meta-character either triggers a state transition, or makes the regex string untranslatable.
+ */
+auto normal_state_transition(
+        TranslatorState& state,
+        string_view::const_iterator& it,
+        string& wildcard_str,
+        [[maybe_unused]] RegexToWildcardTranslatorConfig const& config
+) -> error_code {
+    auto const ch{*it};
+    switch (ch) {
+        case '.':
+            state.set_next_state(TranslatorState::RegexPatternState::DOT);
+            break;
+        case cRegexEndAnchor:
+            state.set_next_state(TranslatorState::RegexPatternState::END);
+            break;
+        case cRegexZeroOrMore:
+            return ErrorCode::Star;
+        case cRegexOneOrMore:
+            return ErrorCode::Plus;
+        case cRegexZeroOrOne:
+            return ErrorCode::Question;
+        case '|':
+            return ErrorCode::Pipe;
+        case cRegexStartAnchor:
+            return ErrorCode::Caret;
+        case ')':
+            return ErrorCode::UnmatchedParenthesis;
+        default:
+            wildcard_str += ch;
+            break;
+    }
+    return ErrorCode::Success;
+}
+
+/**
+ * Attempt to translate regex wildcard patterns that start with `.` character.
+ *
+ * Performs the following translation if possible:
+ * <ul>
+ *   <li> `.*` gets translated into `*`</li>
+ *   <li> `.+` gets translated into `?*`</li>
+ *   <li> `.` gets translated into `?`</li>
+ * </ul>
+ */
+auto dot_state_transition(
+        TranslatorState& state,
+        string_view::const_iterator& it,
+        string& wildcard_str,
+        [[maybe_unused]] RegexToWildcardTranslatorConfig const& config
+) -> error_code {
+    switch (*it) {
+        case cZeroOrMoreCharsWildcard:
+            wildcard_str += cZeroOrMoreCharsWildcard;
+            break;
+        case cRegexOneOrMore:
+            wildcard_str = wildcard_str + cSingleCharWildcard + cZeroOrMoreCharsWildcard;
+            break;
+        default:
+            wildcard_str += cSingleCharWildcard;
+            // Backtrack the scan by one position to handle the current char in the next iteration.
+            --it;
+            break;
+    }
+    state.set_next_state(TranslatorState::RegexPatternState::NORMAL);
+    return ErrorCode::Success;
+}
+
+// Once we've seen the end anchor, we should not expect any other character to appear.
```

Comment:
Lets just make this a doc string to match everything else. Remember (from guidelines):
> The description should be written in present tense (not in the form of a command). A simple way to remember this is to imagine the description is preceded by “This function ….”.